# Проект по SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.


Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

Задачи:  

 - **проанализировать базу данных**. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.
 - **сформулировать ценностное предложение** для нового продукта.

# Описание данных
Таблица **books**

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

Таблица **authors**

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

Таблица **publishers**

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

Таблица **ratings**

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

Таблица **reviews**

Содержит данные о пользовательских обзорах на книги:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

## Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine


# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])


# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Исследование таблиц

### Таблица books

In [2]:
# посмотрим таблицу books
query = ''' 

SELECT *
FROM books
LIMIT 5 

'''
books = pd.io.sql.read_sql(query, con = engine) 
display(books)
books.dtypes

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

book_id              int64
author_id            int64
title               object
num_pages            int64
publication_date    object
publisher_id         int64
dtype: object

В таблице books видим 6 колонок: `book_id` — идентификатор книги; `author_id` — идентификатор автора; `title` — название книги; `num_pages` — количество страниц; `publication_date` — дата публикации книги; `publisher_id` — идентификатор издателя.

### Таблица authors

In [3]:
# посмотрим таблицу authors
query = ''' 

SELECT *
FROM authors
LIMIT 5 

'''
authors = pd.io.sql.read_sql(query, con = engine) 
display(authors)
authors.dtypes

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

author_id     int64
author       object
dtype: object

В таблице authors присутствуют колонки `publisher_id` — идентификатор издательства; `publisher` — название издательства.

### Таблица publishers

In [4]:
# посмотрим таблицу publishers
query = ''' 

SELECT *
FROM publishers
LIMIT 5 

'''
publishers = pd.io.sql.read_sql(query, con = engine) 
display(publishers)
publishers.dtypes

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

publisher_id     int64
publisher       object
dtype: object

В таблице 2 столбца: `publisher_id` — идентификатор издательства; `publisher` — название издательства;

### Таблица ratings

In [5]:
# посмотрим таблицу ratings
query = ''' 

SELECT *
FROM ratings
LIMIT 5 

'''
ratings = pd.io.sql.read_sql(query, con = engine) 
display(ratings)
ratings.dtypes

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

rating_id     int64
book_id       int64
username     object
rating        int64
dtype: object

В таблице столбцы - `rating_id` — идентификатор оценки; `book_id` — идентификатор книги; `username` — имя пользователя, оставившего оценку; `rating` — оценка книги.

### Таблица reviews

In [6]:
# посмотрим таблицу reviews
query = ''' 

SELECT *
FROM reviews
LIMIT 5 

'''
reviews = pd.io.sql.read_sql(query, con = engine) 
display(reviews)
reviews.dtypes

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

review_id     int64
book_id       int64
username     object
text         object
dtype: object

В таблице 4 колонки:  `review_id` — идентификатор обзора; `book_id` — идентификатор книги; `username` — имя пользователя, написавшего обзор; `text` — текст обзора.

## Задания

### Cколько книг вышло после 1 января 2000 года

In [7]:
query = ''' 

SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01' 

'''   
books_count = pd.io.sql.read_sql(query, con = engine) 
books_count

count
0    819

После 1 января 2000 года вышло **819** книг.

### Количество обзоров и средняя оценка для каждой книги

In [8]:
query = ''' 

SELECT books.title,
       COUNT (distinct review_id) as review_count,
       AVG(rating) AS avg_rating
FROM books
LEFT JOIN reviews ON reviews.book_id = books.book_id
LEFT JOIN ratings ON ratings.book_id = books.book_id

GROUP BY books.book_id
ORDER BY COUNT(reviews.review_id) DESC

'''
reviews_and_avg_rating = pd.io.sql.read_sql(query, con = engine)
reviews_and_avg_rating

title  review_count  \
0                              Twilight (Twilight  #1)             7   
1                  The Hobbit  or There and Back Again             6   
2                               The Catcher in the Rye             6   
3    Harry Potter and the Prisoner of Azkaban (Harr...             6   
4    Harry Potter and the Chamber of Secrets (Harry...             6   
..                                                 ...           ...   
995                          Essential Tales and Poems             0   
996                               Leonardo's Notebooks             0   
997   The Cat in the Hat and Other Dr. Seuss Favorites             0   
998    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
999                            The Natural Way to Draw             0   

     avg_rating  
0      3.662500  
1      4.125000  
2      3.825581  
3      4.414634  
4      4.287500  
..          ...  
995    4.000000  
996    4.000000  
997    5.000000  
998    3.666667  
999    3.000000  

[1000 rows x 3 columns]

Видим, максиманое количество обзоров 7 у книги "Сумерки", Далее наблюдаем Хоббита, "Над пропастью во ржи" и книги про Гарри Поттера с количеством обзоров 6. Также видим средние оценки для каждой книги.

### Издательства, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
query = ''' 

SELECT publishers.publisher,

        COUNT(num_pages) as count_books
        
FROM books
INNER JOIN publishers on publishers.publisher_id = books.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY count_books DESC 
LIMIT 3

'''

publishers_big_books = pd.io.sql.read_sql(query, con = engine)
publishers_big_books

publisher  count_books
0             Penguin Books           42
1                   Vintage           31
2  Grand Central Publishing           25

Издательства, выпускающие больше всего книг толще 50 страниц, - это `Penguin Books`, `Vintage` и `Grand Central Publishing`.

### Автор с самой высокой средней оценкой книг (только книги с 50 и более оценками)

In [10]:
query = ''' 

SELECT authors.author,
        AVG(ratings.rating) as mean_rating
FROM books
INNER JOIN authors on authors.author_id = books.author_id
INNER JOIN ratings on ratings.book_id = books.book_id
WHERE books.book_id IN (SELECT book_id
                        FROM ratings
                        GROUP BY book_id
                        HAVING COUNT(rating) >= 50)
GROUP BY authors.author
ORDER BY mean_rating DESC
LIMIT 1 


'''    
top_author = pd.io.sql.read_sql(query, con = engine)
top_author

author  mean_rating
0  J.K. Rowling/Mary GrandPré     4.287097

Автор с самой высокой средней оценкой книг среди книг с 50 и более оценками - J.K. Rowling/Mary GrandPré.

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [11]:
query = ''' 


SELECT SUM(group_table.count_rev)/COUNT(group_table.count_rev) as mean
        FROM (SELECT reviews.username, 
        COUNT(reviews.username) AS count_rev
FROM reviews
WHERE reviews.username IN (SELECT ratings.username
                            FROM ratings
                            GROUP BY ratings.username
                            HAVING COUNT(ratings.rating_id) > 50)
                  
                            GROUP BY username) AS group_table 
                            

'''         
review_avg_count = pd.io.sql.read_sql(query, con = engine)
review_avg_count

mean
0  24.333333

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - **24**.

## Вывод

По итогам анализа было выведено и изучено 5 таблиц из базы данных.

Было выявлено:

1) После 1 января 2000 года вышло **819** книг.

2) Максиманое количество обзоров 7 у книги "Сумерки", Далее наблюдаем Хоббита, "Над пропастью во ржи" и книги про Гарри Поттера с количеством обзоров 6. Также видим средние оценки для каждой книги.

3) Издательства, выпускающие больше всего книг толще 50 страниц, - это Penguin Books, Vintage и Grand Central Publishing.

4) Автор с самой высокой средней оценкой книг среди книг с 50 и более оценками - J.K. Rowling/Mary GrandPré.\

5) Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.

По итогам анализа, видно, что пользователи охотно читают фантастику. При этом, они активно оставляют отзывы и оценивают книги. Возможно, хорошей идеей было бы увеличить разнообразие книг этого жанра, так как в пандемию людям все чаще хочется уходить от реальности. Также можно было бы поработать над смежными жанрами, типа детектив-фантастика и проч.